### ISSUE TITLE
`reportUnusedParameter` false positive on overloads that are decorated with a class decorator

### ISSUE THREAD

**Issue by DetachHead** [DetachHead/basedpyright#604]
Code sample in [basedpyright playground](https://basedpyright.com/?pythonVersion=3.13&typeCheckingMode=all&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQDClZZlARmcQFBcDGrAZwFQAYmDABtAFRSACvQAqAXQBcPKBqjViwKAH09qJDAMAKAcTLB6wAK4peMJGBQrGzVh2IT5UZQEo1TWCoADpw9U1tXQNeDzMLK3opShA0ATdZUNT05KkAawB3agyoLKKcgX8oAFoAPj8gkI1w0J5%2BSiEoACFUpo0AATEwLmCBolIKGlGonShgcXNLayhO6mA3VBhq%2BqgBGBAVVqgAYigQYhw8AFUUWwtqWVTKCGIYUhnB4c%2BocZJyKjUH7ReaLRIrNYbPYHHYNLZHcKnc6XXAwW73YiPZ6vd4gSIaEELMBLJKrATrNxgNgAK2IjlhBBpdJg-RCCBAW1MkP8QA)

```python
from typing import overload, Callable

class Foo[**P, T]:

    def __init__(self, function: Callable[P, T]):
        ...

    def __call__(self, *args: P.args, **kwds: P.kwargs) -> T:
        ...

class Bar:
    @Foo
    @overload
    def foo(self, asdf: int) -> str:... # reportUnusedParameter
    @Foo
    @overload
    def foo(self, asdf: str) -> int:... # reportUnusedParameter

    def foo(self, asdf: object) -> object:
        print(asdf)
```

the same issue also occurs [upstream](https://pyright-play.net/?code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgDRQDClZZlARmcQFBcDGrAZwFQAYmDABtAFRSACvQAqAXQBcPKBqjViwKAH09qJDAMAKAcTLB6wAK4peMJGBQrGzVh2IT5UZQEo1TWCoADpw9U1tXQNeDzMLK3opShA0ATdZUNT05KkAawB3agyoLKKcgX8oAFoAPj8gkI1w0J5%2BSiEoACFUpo0AATEwLmCBolIKGlGonShgcXNLayhO6mA3VBhq%2BqgBGBAVVpnB4ZOocZJyKmpz6PnFxJW1jb2DnYato4i7uYWwJZJVYCdZuMBsABWxEcHwIkOhMH6IQQIC2phe-iAA), but it's more of an issue in basedpyright because we support reporting it as an error instead of just an IDE hint
